grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java

```diff
@@ -0,0 +1,323 @@
+
+/**
+ * A {@link LoadBalancer} that uses the GRPCLB protocol.
+ */
+class GrpclbLoadBalancer extends LoadBalancer {
+  private static final Logger logger = Logger.getLogger(GrpclbLoadBalancer.class.getName());
+
+  private final Object lock = new Object();
+  private final String serviceName;
+  private final TransportManager tm;
+
+  // General states
+  @GuardedBy("lock")
+  private final BlankFutureProvider<ClientTransport> pendingPicks =
+      new BlankFutureProvider<ClientTransport>();
+  @GuardedBy("lock")
+  private Throwable lastError;
+
+  @GuardedBy("lock")
+  private boolean closed;
+
+  // Load-balancer service states
+  @GuardedBy("lock")
+  private EquivalentAddressGroup lbAddresses;
+  @GuardedBy("lock")
+  private ClientTransport lbTransport;
+  @GuardedBy("lock")
+  private StreamObserver<LoadBalanceResponse> lbResponseObserver;
+  @GuardedBy("lock")
+  private StreamObserver<LoadBalanceRequest> lbRequestWriter;
+
+  // Server list states
+  @GuardedBy("lock")
+  private HashMap<SocketAddress, ResolvedServerInfo> servers;
+  @GuardedBy("lock")
+  @VisibleForTesting
+  private RoundRobinServerList roundRobinServerList;
+
+  private ExecutorService executor;
+  private ScheduledExecutorService deadlineCancellationExecutor;
+
+  GrpclbLoadBalancer(String serviceName, TransportManager tm) {
+    this.serviceName = serviceName;
+    this.tm = tm;
+    executor = SharedResourceHolder.get(GrpcUtil.SHARED_CHANNEL_EXECUTOR);
+    deadlineCancellationExecutor = SharedResourceHolder.get(GrpcUtil.TIMER_SERVICE);
+  }
+
+  @VisibleForTesting
+  StreamObserver<LoadBalanceResponse> getLbResponseObserver() {
+    synchronized (lock) {
+      return lbResponseObserver;
+    }
+  }
+
+  @VisibleForTesting
+  RoundRobinServerList getRoundRobinServerList() {
+    synchronized (lock) {
+      return roundRobinServerList;
+    }
+  }
+
+  @Override
+  public ListenableFuture<ClientTransport> pickTransport(@Nullable RequestKey requestKey) {
+    RoundRobinServerList serverListCopy;
+    synchronized (lock) {
+      Preconditions.checkState(!closed, "already closed");
+      if (roundRobinServerList == null) {
+        if (lastError == null) {
+          return pendingPicks.newBlankFuture();
+        } else {
+          return Futures.immediateFailedFuture(lastError);
+        }
+      }
+      serverListCopy = roundRobinServerList;
+    }
+    return serverListCopy.getTransportForNextServer();
+  }
+
+  @Override
+  public void handleResolvedAddresses(
+      List<ResolvedServerInfo> updatedServers, Attributes config) {
+    synchronized (lock) {
+      ArrayList<SocketAddress> addrs = new ArrayList<SocketAddress>(updatedServers.size());
+      for (ResolvedServerInfo serverInfo : updatedServers) {
+        addrs.add(serverInfo.getAddress());
+      }
+      EquivalentAddressGroup newLbAddresses = new EquivalentAddressGroup(addrs);
+      if (!newLbAddresses.equals(lbAddresses)) {
+        lbAddresses = newLbAddresses;
```

Comment:
I changed `TransportManager.updateRetainedTransports()` to accept a collection of `EquivalentAddressGroup`s, which is called whenever `lbAddresses` or `servers` changes, passing a merged collection.